### PR TITLE
docs: add asset encoding specs to README

### PR DESCRIPTION
**Pull request by miltonjonat**
The main motivation for this PR is the lack of information about how to decode deposit inputs, which currently requires developers to inspect the Solidity source code.

Other than that, this PR is intended to help structure the official site documentation in regards to explaining how asset handling operations are done.

The PR includes the following additions to the main README:
- Input payload encodings for deposits made by the portals
- Better explanation of how voucher payloads should be encoded

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -106,6 +106,30 @@ Currently, we support the following types of assets:
 - [ERC-721](https://ethereum.org/en/developers/docs/standards/tokens/erc-721/) (Non-fungible tokens)
 - [ERC-1155](https://ethereum.org/en/developers/docs/standards/tokens/erc-1155/) (Multi-tokens)
 
+### Input encodings for deposits
+
+As explained above, when a portal teleports assets to the execution layer, it adds an input to the DApp's input box, which will then need to be interpreted and validated by the off-chain machine. In order to do that, the machine will need to understand and decode the input's payload.
```
Comment (guidanoli) [inline README.md:111]: I would change the phrasing here a bit. It seems like the addition of an input is simply a side effect of some magical operation, which is the teleportation of assets. Rather, the teleportation is only possible with the addition of an input. So, I think it would be more accurate if we said something like... **in order** for the portal to teleport assets from the base layer to the execution layer, it adds an input to the DApp's input box, [...].

File: README.md [hunk 2]
```diff
@@ -106,6 +106,30 @@ Currently, we support the following types of assets:
...
+
+| Asset | Payload |
+| :- | :- |
+| Ether | `address sender, uint256 value, bytes execLayerData` |
+| ERC-20 | `bool success, address token, address sender, uint256 amount, bytes execLayerData` |
+| ERC-721 | `address token, address sender, uint256 tokenId, bytes data` `*` |
+| ERC-1155 (single) | `address token, address sender, uint256 tokenId, uint256 value, bytes data` `*`|
+| ERC-1155 (batch) | `address token, address sender, bytes data` `**` |
+
+**`*`** The ERC-721 and ERC-1155 (single) `data` fields correspond to the ABI-encoding of: `bytes baseLayerData, bytes execLayerData`
```
Comment (guidanoli) [inline README.md:123]: I think it's healthy to highlight here that for these `data` fields, are using the **standard** ABI encoding, and not the packed one.

File: README.md [hunk 3]
```diff
@@ -106,6 +106,30 @@ Currently, we support the following types of assets:
...
+
+**`**`** The ERC-1155 (batch) `data` field corresponds to the ABI-encoding of: `uint256[] tokenIds, uint256[] values, bytes baseLayerData, bytes execLayerData`
+
+As an example, the deposit of 100 wei (of Ether) sent by address `0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266` with data `0xabcd` would result in the following input payload:
```
Comment (guidanoli) [inline README.md:127]: Wei is generally written capitalized.

File: README.md [hunk 4]
```diff
@@ -124,6 +148,14 @@ Because of their generality, vouchers can be used in a wide range of application
 | ERC-1155 | Token contract | `safeTransferFrom(address,address,uint256,uint256,data)` [:page_facing_up:](https://eips.ethereum.org/EIPS/eip-1155#specification) |
 | ERC-1155 | Token contract | `safeBatchTransferFrom(address,address,uint256[],uint256[],data)` [:page_facing_up:](https://eips.ethereum.org/EIPS/eip-1155#specification) [^3] |
 
+Please note that the voucher payload should be encoded according to the [Ethereum ABI specification for calling contract functions](https://docs.soliditylang.org/en/v0.8.19/abi-spec.html). As such, it should start with the first four bytes of the Keccak-256 hash of the function selector string (as given in the table above), followed by the ABI-encoded parameter values.
```
Comment (guidanoli) [inline README.md:151]: The function **selector** is the first four bytes of the Keccak-256 hash of the function **signature**. For example, the EIP-20 transfer function has signature `transfer(address,uint256)` and selector `0xa9059cbb`.

File: README.md [hunk 5]
```diff
@@ -124,6 +148,14 @@ Because of their generality, vouchers can be used in a wide range of application
...
+
+As an example, the voucher for a simple ERC-20 transfer (2nd line in the table above) to address `0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266` with amount 100 should specify the following payload:
```
Comment (guidanoli) [inline README.md:153]: We could rephrase this to "a simple ERC-20 transfer (...) of 100 tokens to address `...` should [...]"

File: README.md [hunk 6]
```diff
@@ -106,6 +106,30 @@ Currently, we support the following types of assets:
...
+
+The input payloads for deposits are always specified as packed ABI-encoded parameters, as detailed below:
+
+| Asset | Payload |
+| :- | :- |
+| Ether | `address sender, uint256 value, bytes execLayerData` |
+| ERC-20 | `bool success, address token, address sender, uint256 amount, bytes execLayerData` |
+| ERC-721 | `address token, address sender, uint256 tokenId, bytes data` `*` |
+| ERC-1155 (single) | `address token, address sender, uint256 tokenId, uint256 value, bytes data` `*`|
+| ERC-1155 (batch) | `address token, address sender, bytes data` `**` |
```
Comment (guidanoli) [inline README.md:121]: I feel like we need to explain each component of each input, but this table would turn too wide. Maybe we could try exploding it into several subsections. This will give more space for more detailed descriptions.

Comment (miltonjonat) [reply README.md:121]: I see what you mean, but I think it's more or less understandable as such. As you said, the table is already too big, and I also think that adding more subsections will be a little too much for a README, it's already really long. In any case, the meaning/description of each component is actually already included in the website API documentation, since they all derive from the arguments used when calling the deposit methods.

File: README.md (hunk 3 above)
Comment (miltonjonat) [reply README.md]: Done in https://github.com/cartesi/rollups/pull/42/commits/e4ef2ae0655fe2b21af3afedda7828fbdb4aab80

File: README.md (hunk 4 above)
Comment (miltonjonat) [reply README.md]: Done in https://github.com/cartesi/rollups/pull/42/commits/e4ef2ae0655fe2b21af3afedda7828fbdb4aab80

File: README.md (hunk 5 above)
Comment (guidanoli) [reply README.md:153]: I see your point now. Thanks for the clarification! :-)

File: README.md (hunk 6 above)
Comment (guidanoli) [reply README.md:121]: Ok. I'm fine with this README being just a superficial documentation, while having a much thorough documentation available elsewhere. :-)

File: README.md (hunk 5 above)
Comment (miltonjonat) [reply README.md:153]: Oh I was hoping you would have a great suggestion as to how to express this better!

**Review by gligneul — APPROVED** [review]

File: README.md (hunk 5 above)
Comment (guidanoli) [reply README.md:153]: Well, clearly the phrase "Bob has $x$ tokens" is ambiguous, as our interpretations have diverged. Does it mean $x$ or $x / 10^D$ [^1]?

**In my interpretation**, if you say Bob has 42 tokens, then I will think that if you call `balanceOf(bob)`, then you'll get the `uint256` value of 42 exactly. Now, for the user, this number may be displayed differently depending on the number of decimal places. So, if token FOO uses, let's say, 2 decimal places, then it would appear to me as 0.42 FOO, but in reality I have 42 tokens.

**Another interpretation** (which is probably that of most users) is that `balanceOf(bob)` would return $42*10^D$. In this case, are users even aware they are actually dealing with discreet tokens, and that EIP-20 doesn't deal with floating point numbers? I don't know if they have a name for the discreet tokens. If so, we should use this name.

[^1]: where D is the number of decimal places

**Review by guidanoli — APPROVED** [review]

File: README.md (hunk 5 above)
Comment (pedroargento) [reply README.md:153]: I think the correct way of saying is 42 tokens or 42e18 amount. Just like if you say you are transfering 42 ethers you .

So if bob has 42 tokens, balanceOf should return 42e18.

I would write something "a voucher for 100 tokens would have as parameters (...) 100e18" with a footnote explaining that the parameter is in the undivisible unit.

Comment (miltonjonat) [reply README.md:153]: Thanks for the inputs guys. And all because I want a small number (0x64) to appear in the payload! Since this PR was approved I am closing it as is, we can improve later on.
